Adjust the length of blob cache docs for Lucene metadata files

tlrx opened the pull request:
Today searchable snapshots `IndexInput` implementations use the _blob store cache_ to cache the first `4096` bytes of every Lucene files. After some experiments we think that we could adjust the length of the cached data depending of the Lucene file that is read, caching up to 64KB for Lucene _metadata_ files (ie files that are fully read when a Directory is opened) and only 1KB for other files.

The files that are cached up to 64KB are the files with the following extensions:

>         "cfe", // compound file's entry table
>         "dvm", // doc values metadata file
>         "fdm", // stored fields metadata file
>         "fnm", // field names metadata file
>         "kdm", // Lucene 8.6 point format metadata file
>         "nvm", // norms metadata file
>         "tmd", // Lucene 8.6 terms metadata file
>         "tvm", // terms vectors metadata file
>         "vem"  // Lucene 9.0 indexed vectors metadata

The 64KB limit can be configured on a per index basis  through a new index setting.

This change is extracted from #69283 and does not address the caching of CFS files.

tlrx committed:
Adjust the length of blob cache docs for Lucene metadata files

elasticmachine commented:
Pinging @elastic/es-distributed (Team:Distributed)

tlrx commented on x-pack/plugin/searchable-snapshots/src/internalClusterTest/java/org/elasticsearch/blobstore/cache/SearchableSnapshotsBlobStoreCacheIntegTests.java:
```diff
@@ -67,6 +77,35 @@
+        builder.put(SnapshotsService.SNAPSHOT_CACHE_REGION_SIZE_SETTING.getKey(), blobCacheMaxLength);
+        builder.put(SnapshotsService.SHARED_CACHE_RANGE_SIZE_SETTING.getKey(), blobCacheMaxLength);
+        builder.put(FrozenCacheService.FROZEN_CACHE_RECOVERY_RANGE_SIZE_SETTING.getKey(), blobCacheMaxLength);
+        cacheSettings = builder.build();
```
I'm trying to randomize these settings.

tlrx commented on x-pack/plugin/searchable-snapshots/src/internalClusterTest/java/org/elasticsearch/blobstore/cache/SearchableSnapshotsBlobStoreCacheIntegTests.java:
```diff
@@ -79,28 +118,25 @@ protected int numberOfReplicas() {
+                builder.startObject().field("text", randomRealisticUnicodeOfCodepointLengthBetween(5, 50)).field("num", i).endObject();
+                indexRequestBuilders.add(client().prepareIndex(indexName).setSource(builder));
+            }
+            indexRandom(true, true, true, indexRequestBuilders);
```
It now uses `dummyDocs` in order to generate deletions

tlrx commented on x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/blobstore/cache/BlobStoreCacheService.java:
```diff
@@ -155,4 +159,46 @@ public void onFailure(Exception e) {
+        if (METADATA_FILES_EXTENSIONS.contains(fileExtension)) {
+            final long maxAllowedLengthInBytes = maxMetadataLength.getBytes();
+            if (fileLength > maxAllowedLengthInBytes) {
+                logger.warn(
```
Maybe too verbose?

tlrx committed:
revert cache range to 4KB

ywelsch commented on x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/SearchableSnapshots.java:
```diff
@@ -185,6 +187,32 @@
+            }
+        },
+        Setting.Property.IndexScope,
+        Setting.Property.NodeScope,
```
Let's make it purely indexScope?

tlrx replied:
Good catch, I pushed b5f2463ba4c80ddf4d87d969e969aced2d70f849

ywelsch commented on x-pack/plugin/searchable-snapshots/src/internalClusterTest/java/org/elasticsearch/blobstore/cache/SearchableSnapshotsBlobStoreCacheIntegTests.java:
```diff
@@ -67,6 +77,35 @@
+
+        // Frozen (shared cache) cache should be large enough to not cause direct reads
+        builder.put(SnapshotsService.SNAPSHOT_CACHE_SIZE_SETTING.getKey(), ByteSizeValue.ofMb(128));
+        // Align ranges to match the blob cache max length
```
why have this alignment?

ywelsch reviewed (commented):
Thanks for opening this PR. I've done a first pass and left some comments.

tlrx committed:
remove cfe

tlrx committed:
add other types with assertions

tlrx committed:
log only per hour

tlrx committed:
remove comment

tlrx committed:
remove assert on cached blob size

tlrx committed:
reword TODO

tlrx committed:
remove NodeScope

tlrx committed:
Assume changed docs range as cache miss

tlrx committed:
fix SearchableSnapshotDirectoryStatsTests

tlrx committed:
Merge branch 'master' into adjust-cached-blob-size-for-metadata-files

tlrx commented on x-pack/plugin/searchable-snapshots/src/internalClusterTest/java/org/elasticsearch/blobstore/cache/SearchableSnapshotsBlobStoreCacheIntegTests.java:
```diff
@@ -163,106 +209,98 @@ public void testBlobStoreCache() throws Exception {
-                if (indexInputStats.getTotalSize() <= BlobStoreCacheService.DEFAULT_CACHED_BLOB_SIZE * 2
-                    || mayReadMoreThanHeader == false) {
+                final boolean mayReadMoreThanHeader = indexInputStats.getFileExt().equals("cfs")
+                    || indexInputStats.getFileExt().equals("cfe");
```
It's a leftover from debugging sessions, I removed them in 1d9cda0d11f175f803b517a8032092548b1ba6d3

tlrx commented on x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/blobstore/cache/BlobStoreCacheService.java:
```diff
@@ -155,4 +159,46 @@ public void onFailure(Exception e) {
         }
     }
+
+    private static final Set<String> METADATA_FILES_EXTENSIONS = Set.of(
```
Sure, I pushed 40f7557a42a0123ad1e322241279b9ce61e5e768

tlrx commented on x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/cache/CachedBlobContainerIndexInput.java:
```diff
@@ -162,25 +163,18 @@ protected void doReadInternal(ByteBuffer b) throws IOException {
 
             // We try to use the cache index if:
             // - the file is small enough to be fully cached
```
Addressed in 0cc9e9d428cd94926e786e0605ad697174eeff2e

tlrx commented on x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/cache/CachedBlobContainerIndexInput.java:
```diff
@@ -274,11 +268,12 @@ protected void doReadInternal(ByteBuffer b) throws IOException {
                     // - use an intermediate copy buffer to read the file in sensibly-sized chunks
                     // - release the buffer once the indexing operation is complete
-                    assert indexCacheMissLength <= COPY_BUFFER_SIZE : indexCacheMiss;
+                    // assert indexCacheMissLength <= COPY_BUFFER_SIZE : indexCacheMiss;
```
Sure -> 6e70ff3a0c057975add184a6ce654b5e3687e38f

tlrx commented on x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/cache/CachedBlobContainerIndexInput.java:
```diff
@@ -274,11 +268,12 @@ protected void doReadInternal(ByteBuffer b) throws IOException {
                 final Future<Integer> readFuture = cacheFile.readIfAvailableOrPending(indexCacheMiss, channel -> {
                     final int indexCacheMissLength = toIntBytes(indexCacheMiss.length());
 
+                    // TODO: should we use BigArrays?
```
Ok - 6e70ff3a0c057975add184a6ce654b5e3687e38f

tlrx commented on x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/cache/CachedBlobContainerIndexInput.java:
```diff
@@ -600,7 +595,8 @@ public IndexInput slice(String sliceDescription, long offset, long length) {
             defaultRangeSize,
-            recoveryRangeSize
+            recoveryRangeSize,
+            ByteRange.EMPTY // TODO implement blob cache for CFS
```
Yes, I reworded this TODO in 5d28df834b81a0823abda9abf60944647ad8c195 but it should be adressed in a follow up.

tlrx committed:
Merge branch 'master' into adjust-cached-blob-size-for-metadata-files

tlrx commented on x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/blobstore/cache/BlobStoreCacheService.java:
```diff
@@ -41,7 +45,7 @@
     private static final Logger logger = LogManager.getLogger(BlobStoreCacheService.class);
 
-    public static final int DEFAULT_CACHED_BLOB_SIZE = ByteSizeUnit.KB.toIntBytes(4);
+    public static final int DEFAULT_CACHED_BLOB_SIZE = ByteSizeUnit.KB.toIntBytes(1);
```
Sorry for the delayed response, I had to deal with other duties.

Looking at the current code in 7.x, reducing the size of cached blob might throw an exception in Cached/FrozenIndexInput at this line:

> final BytesRefIterator cachedBytesIterator = cachedBlob.bytes().slice(toIntBytes(position), length).iterator();

in the case `position` is larger than one or two buffered reads.

I suggest 
- to keep the 4KB/8KB limit for non metadata files as long as there is at least a node in version < 7.13 in the cluster
- to write some BWC test that ensure indices are correctly assigned during rolling upgrades

tlrx committed:
Merge branch 'master' into adjust-cached-blob-size-for-metadata-files

tlrx committed:
Add BWC support

tlrx committed:
Merge branch 'master' into adjust-cached-blob-size-for-metadata-files

tlrx committed:
Add BWC tests

tlrx committed:
Merge branch 'master' into adjust-cached-blob-size-for-metadata-files

tlrx committed:
rename and fix tests (hopefully)

tlrx committed:
I hate you spotless

tlrx commented:
Thanks @ywelsch for your review. I updated the code according to your comments. 

Concerning https://github.com/elastic/elasticsearch/pull/69431#discussion_r581009695 and the behavior with previous versions, I followed my suggestion in https://github.com/elastic/elasticsearch/pull/69431#discussion_r582646610 and I added a rolling upgrade test named `SearchableSnapshotsRollingUpgradeIT`.

This integration test is composed of 4 tests: 
- two are tests that mount a snapshot as an index in the previous version and verifies that the index is recovered and searchable on mixed and upgraded version.
- the two other tests are a bit tricky and try to test the behavior of the blob cache when cached blob documents are created from an old version or a new version. Those tests only runs in mixed version cluster. I validated that it throws the exception I suspected in https://github.com/elastic/elasticsearch/pull/69431#discussion_r582646610.

This is ready for another review.

ywelsch reviewed (approved):
LGTM. Thanks Tanguy!

tlrx commented:
Thanks Yannick! 

I'm going to backport this down to 7.12 and adjusts the BWC version OLD_CACHED_BLOB_SIZE_VERSION in follow up PRs.
